tfe/plugin_provider.go

```diff
@@ -0,0 +1,252 @@
+						Description: descriptions["hostname"],
+						Optional:    true,
+					},
+					&tfprotov5.SchemaAttribute{
+						Name:        "token",
+						Type:        tftypes.String,
+						Description: descriptions["token"],
+						Optional:    true,
+					},
+					&tfprotov5.SchemaAttribute{
+						Name:        "ssl_skip_verify",
+						Type:        tftypes.Bool,
+						Description: descriptions["ssl_skip_verify"],
+						Optional:    true,
+					},
+				},
+			},
+		},
+		dataSourceSchemas: map[string]*tfprotov5.Schema{
+			"tfe_state_outputs": {
+				Version: 1,
+				Block: &tfprotov5.SchemaBlock{
+					Version: 1,
+					Attributes: []*tfprotov5.SchemaAttribute{
+						{
+							Name:     "id",
+							Type:     tftypes.String,
+							Computed: true,
+						},
+						{
+							Name:            "workspace",
+							Type:            tftypes.String,
+							Description:     "The workspace to fetch the remote state from.",
+							DescriptionKind: tfprotov5.StringKindPlain,
+							Required:        true,
+						},
+						{
+							Name:            "organization",
+							Type:            tftypes.String,
+							Description:     "The organization to fetch the remote state from.",
+							DescriptionKind: tfprotov5.StringKindPlain,
+							Required:        true,
+						},
+						{
+							Name:     "values",
+							Type:     tftypes.DynamicPseudoType,
+							Optional: true,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+		dataSourceRouter: map[string]func(p *pluginProviderServer) tfprotov5.DataSourceServer{
+			"tfe_state_outputs": newDataSourceStateOutputs,
+		},
+	}
+}
+
+func retrieveProviderMeta(req *tfprotov5.ConfigureProviderRequest) (providerMeta, error) {
+	meta := providerMeta{}
+	config := req.Config
+	val, err := config.Unmarshal(tftypes.Object{
+		AttributeTypes: map[string]tftypes.Type{
+			"hostname":        tftypes.String,
+			"token":           tftypes.String,
+			"ssl_skip_verify": tftypes.Bool,
+		}})
+
+	if err != nil {
+		return meta, fmt.Errorf("Could not unmarshal ConfigureProviderRequest %v", err)
+	}
+	var hostname string
+	var token string
+	var valMap map[string]tftypes.Value
+	err = val.As(&valMap)
+	if err != nil {
+		return meta, fmt.Errorf("Could not set the schema attributes to map %v", err)
+	}
+	err = valMap["hostname"].As(&hostname)
+	if err != nil {
+		return meta, fmt.Errorf("Could not set the hostname value to string %v", err)
+	}
+	if hostname == "" && os.Getenv("TFE_HOSTNAME") == "" {
+		return meta, fmt.Errorf("Hostname must not be empty")
+	} else if hostname == "" && os.Getenv("TFE_HOSTNAME") != "" {
+		hostname = os.Getenv("TFE_HOSTNAME")
+	}
+
+	err = valMap["token"].As(&token)
+	if err != nil {
+		return meta, fmt.Errorf("Could not set the token value to string %v", err)
+	}
+	if token == "" && os.Getenv("TFE_TOKEN") == "" {
+		return meta, fmt.Errorf("Token must not be empty")
+	} else if token == "" && os.Getenv("TFE_TOKEN") != "" {
+		token = os.Getenv("TFE_TOKEN")
+	}
```

Comment:
You should probably check these aren't null or unknown before trying to call `.As` on the `valMap` values, because they could be and I believe you'd return an error at the moment, and not a particularly helpful one at that.